discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,266 @@
+	c, _, err := zk.Connect(addresses, time.Second, zk.WithEventCallback(cb))
+	return c, err
+}
+
+// NewZooKeeperConfig creates a new service discovery backend for zookeeper
+func NewZooKeeperConfig(raw interface{}) (*ZooKeeper, error) {
+	zookeeper := &ZooKeeper{
+		Prefix:    "/containerpilot",
+		eventChan: make(chan zk.Event),
+	}
+	config := &struct {
+		Address string `mapstructure:"address"`
+	}{}
+	if err := utils.DecodeRaw(raw, &config); err != nil {
+		return nil, err
+	}
+	conn, err := connection([]string{config.Address}, zookeeper.eventChanCallBack)
+	if err != nil {
+		return nil, err
+	}
+	zookeeper.Client = conn
+	return zookeeper, nil
+}
+
+// Deregister removes this instance from the registry
+func (conn *ZooKeeper) Deregister(service *discovery.ServiceDefinition) {
+	key := conn.getNodeKey(service)
+	exists, _, err := conn.Client.Exists(key)
+	if err != nil {
+		log.Warnf("Error on deregistering %s. Node should exists: %s", key, err)
+	}
+	if exists {
+		log.Warnf("Deregistering %s", key)
+		conn.Client.Delete(conn.getNodeKey(service), -1)
+	}
+}
+
+// MarkForMaintenance removes this instance from the registry
+func (conn *ZooKeeper) MarkForMaintenance(service *discovery.ServiceDefinition) {
+	conn.Deregister(service)
+}
+
+// SendHeartbeat refreshes the associated zookeeper node by
+// re-registering it.
+func (conn *ZooKeeper) SendHeartbeat(service *discovery.ServiceDefinition) {
+	if err := conn.registerService(service); err != nil {
+		log.Warnf("Error registering service %s: %s", service.Name, err)
+	}
+}
+
+func (conn *ZooKeeper) parentPath(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, service.Name)
+}
+
+func (conn *ZooKeeper) getNodeKey(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.parentPath(service), service.ID)
+}
+
+func (conn *ZooKeeper) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, appName)
+}
+
+var zookeeperUpstreams = make(map[string][]ServiceNode)
+
+// CheckForUpstreamChanges checks another zookeeper node for changes
+func (conn *ZooKeeper) CheckForUpstreamChanges(backendName, backendTag string) bool {
+	// TODO: is there a way to filter by tag in zookeeper?
+	services, err := conn.getServices(backendName)
+	if err != nil {
+		log.Errorf("Failed to query %v: %s", backendName, err)
+		return false
+	}
+	didChange := zookeeperCompareForChange(zookeeperUpstreams[backendName], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		zookeeperUpstreams[backendName] = services
+	}
+	return didChange
+}
+
+func (conn *ZooKeeper) getServices(appName string) ([]ServiceNode, error) {
+	var services []ServiceNode
+
+	key := conn.getAppKey(appName)
+	children, _, error := conn.Client.Children(key)
+	if error != nil {
```

Comment:
`error` is a keyword, this should be `err`. (I'm _really_ surprised `go vet` doesn't call this out.)